# Project Description
A semiconductor manufacturer has 437 sensors tracking their process and whether. The sensor outputs is associated with a pass or fail value. Using the data from these sensors build a model that can predict whether a semiconductor should pass or fail.

# Data Preprocessing

## Importing the Libraries

In [10]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the Dataset

In [11]:
dataset = pd.read_csv('SemiconductorManufacturingProcessDataset.csv')

## Showing the Dataset in a Table

In [12]:
pd.DataFrame(dataset)

Time  Sensor 1  Sensor 2   Sensor 3   Sensor 4  Sensor 5  \
0      7/19/2008 11:55   3030.93   2564.00  2187.7333  1411.1265    1.3602   
1      7/19/2008 12:32   3095.78   2465.14  2230.4222  1463.6606    0.8294   
2      7/19/2008 13:17   2932.61   2559.94  2186.4111  1698.0172    1.5102   
3      7/19/2008 14:43   2988.72   2479.90  2199.0333   909.7926    1.3204   
4      7/19/2008 15:22   3032.24   2502.87  2233.3667  1326.5200    1.5334   
...                ...       ...       ...        ...        ...       ...   
1562  10/16/2008 15:13   2899.41   2464.36  2179.7333  3085.3781    1.4843   
1563  10/16/2008 20:49   3052.31   2522.55  2198.5667  1124.6595    0.8763   
1564   10/17/2008 5:26   2978.81   2379.78  2206.3000  1110.4967    0.8236   
1565   10/17/2008 6:01   2894.92   2532.01  2177.0333  1183.7287    1.5726   
1566   10/17/2008 6:07   2944.92   2450.76  2195.4444  2914.1792    1.5978   

      Sensor 6  Sensor 7  Sensor 8  Sensor 9  ...  Sensor 429  Sensor 430  \
0      97.6133    0.1242    1.5005    0.0162  ...     14.9509      0.5005   
1     102.3433    0.1247    1.4966   -0.0005  ...     10.9003      0.5019   
2      95.4878    0.1241    1.4436    0.0041  ...      9.2721      0.4958   
3     104.2367    0.1217    1.4882   -0.0124  ...      8.5831      0.4990   
4     100.3967    0.1235    1.5031   -0.0031  ...     10.9698      0.4800   
...        ...       ...       ...       ...  ...         ...         ...   
1562   82.2467    0.1248    1.3424   -0.0045  ...     11.7256      0.4988   
1563   98.4689    0.1205    1.4333   -0.0061  ...     17.8379      0.4975   
1564   99.4122    0.1208       NaN       NaN  ...     17.7267      0.4987   
1565   98.7978    0.1213    1.4622   -0.0072  ...     19.2104      0.5004   
1566   85.1011    0.1235       NaN       NaN  ...     22.9183      0.4987   

      Sensor 431  Sensor 432  Sensor 433  Sensor 434  Sensor 435  Sensor 436  \
0         0.0118      0.0035      2.3630         NaN         NaN         NaN   
1         0.0223      0.0055      4.4447      0.0096      0.0201      0.0060   
2         0.0157      0.0039      3.1745      0.0584      0.0484      0.0148   
3         0.0103      0.0025      2.0544      0.0202      0.0149      0.0044   
4         0.4766      0.1045     99.3032      0.0202      0.0149      0.0044   
...          ...         ...         ...         ...         ...         ...   
1562      0.0143      0.0039      2.8669      0.0068      0.0138      0.0047   
1563      0.0131      0.0036      2.6238      0.0068      0.0138      0.0047   
1564      0.0153      0.0041      3.0590      0.0197      0.0086      0.0025   
1565      0.0178      0.0038      3.5662      0.0262      0.0245      0.0075   
1566      0.0181      0.0040      3.6275      0.0117      0.0162      0.0045   

      Sensor 437  Pass/Fail  
0            NaN       Pass  
1       208.2045       Pass  
2        82.8602       Fail  
3        73.8432       Pass  
4        73.8432       Pass  
...          ...        ...  
1562    203.1720       Pass  
1563    203.1720       Pass  
1564     43.5231       Pass  
1565     93.4941       Pass  
1566    137.7844       Pass  

[1567 rows x 439 columns]

## Seperate The Input and Output
Here, we put the independent variables in X and the dependent variable in y. 

In [13]:
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, -1].values

## Taking care of missing data

In [14]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
imputer.fit(X[:, 1:])
X[:, 1:] = imputer.transform(X[:, 1:])

## Encoding Categorical Data

### Encoding the Dependent Variable

In [15]:
from sklearn.preprocessing import LabelEncoder
LaEnc = LabelEncoder()
Y = LaEnc.fit_transform(Y)

## Feature Scaling

In [16]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X[:, 1:] = sc.fit_transform(X[:, 1:])

## Remove Specific Columns
Remove column containing date. 

In [17]:
X = np.delete(X, 0, 1)

## Splitting the Dataset into the Training set and Test set

In [18]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 1)

# Classification Model Analysis
## KNN
Train the regression model.

In [19]:
from sklearn.neighbors import KNeighborsClassifier 

# instantiate knn object
knn = KNeighborsClassifier(n_neighbors=3)
# train model
knn = knn.fit(X_train, y_train)

Print out the models accuracy.

In [20]:
knn.score(X_test, y_test)

0.9331210191082803

# Conclusion
The models accuracy using KNN with 3 neighbors is approximately 93%. This could potentially be improved by reducing the number of sensors or by using a different classification algorithm.